File: day18_part1_2020.dart
import 'dart:io';

void main() {
  var file = new File('input.txt');
  var sum = 0;

  file.readAsLines().then((List<String> lines) {
    for (var line in lines) {
      var result = evaluate(line);
      sum += result;
    }
    print(sum);
  });
}

int evaluate(String expression) {
  var tokens = tokenize(expression);
  return evaluateTokens(tokens);
}

List<String> tokenize(String expression) {
  expression = expression.replaceAll("(", "( ");
  expression = expression.replaceAll(")", " )");
  return expression.split(" ");
}

int evaluateTokens(List<String> tokens) {
  var ops = [];
  var vals = [];

  for (var i = 0; i < tokens.length; i++) {
    var token = tokens[i];
    switch (token) {
      case "(":
        ops.add(token);
        break;
      case "+":
      case "*":
        while (ops.length > 0 && ops.last != "(") {
          vals = vals.sublist(0, vals.length - 2)
            ..add(applyOp(ops.last, vals[vals.length - 2], vals.last));
          ops.removeLast();
        }
        ops.add(token);
        break;
      case ")":
        while (ops.last != "(") {
          vals = vals.sublist(0, vals.length - 2)
            ..add(applyOp(ops.last, vals[vals.length - 2], vals.last));
          ops.removeLast();
        }
        ops.removeLast(); // Remove the opening '('
        break;
      default:
        var value = int.parse(token);
        vals.add(value);
    }
  }
  while (ops.length > 0) {
    vals = vals.sublist(0, vals.length - 2)
      ..add(applyOp(ops.last, vals[vals.length - 2], vals.last));
    ops.removeLast();
  }
  return vals[0];
}

int applyOp(String op, int a, int b) {
  switch (op) {
    case "+":
      return a + b;
    case "*":
      return a * b;
    default:
      throw "Unknown operator: $op";
  }
}

File: java/day18_part1_2020.java

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.Deque;

public class Main {

    public static void main(String[] args) {
        try (BufferedReader br = new BufferedReader(new FileReader("input.txt"))) {
            long sum = 0;
            String line;
            while ((line = br.readLine()) != null) {
                sum += evaluate(line);
            }
            System.out.println(sum);
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
        }
    }

    private static long evaluate(String expression) {
        Deque<Character> ops = new ArrayDeque<>();
        Deque<Long> vals = new ArrayDeque<>();
        
        for (int i = 0; i < expression.length(); i++) {
            char token = expression.charAt(i);
            if (token == ' ') continue;
            if (token == '(') {
                ops.push(token);
            } else if (token == '+' || token == '*') {
                while (!ops.isEmpty() && ops.peek() != '(') {
                    vals.push(applyOp(ops.pop(), vals.pop(), vals.pop()));
                }
                ops.push(token);
            } else if (token == ')') {
                while (ops.peek() != '(') {
                    vals.push(applyOp(ops.pop(), vals.pop(), vals.pop()));
                }
                ops.pop();
            } else if (Character.isDigit(token)) {
                long num = 0;
                while (i < expression.length() && Character.isDigit(expression.charAt(i))) {
                    num = num * 10 + (expression.charAt(i) - '0');
                    i++;
                }
                i--;
                vals.push(num);
            }
        }
        while (!ops.isEmpty()) {
            vals.push(applyOp(ops.pop(), vals.pop(), vals.pop()));
        }
        return vals.pop();
    }

    private static long applyOp(char op, long b, long a) {
        if (op == '+') return a + b;
        if (op == '*') return a * b;
        throw new IllegalArgumentException("Unknown operator: " + op);
    }
}
